File: livingrimoire start here/LivinGrimoire swift/LivinGrimoire.swift
//
//  LivinGrimoire.swift
//  LivinGrimoireSwiftV1
//
//  Created by moti barski on 21/06/2022.
//

import Foundation

open class AbsDictionaryDB{
    func save(key:String, value: String){
        // save to DB (override me)
    }
    func load(key:String)->String{
        // override me
        return "null"
    }
}
open class AlgPart{
    var algKillSwitch:Bool = false
    func action(ear: String, skin: String, eye: String) -> String{
        return ""
    }
    func completed() -> Bool{
        //Has finished ?
        return false
    }
    func myName() -> String{
        // Returns the class name
        return String(describing: type(of: self))
                      }
}/*
 it speaks something x times
 a most basic skill.
 also fun to make the chobit say what you want */
class APsay:AlgPart{
    var at:Int=10
    var param:String="hmm"
    convenience init(repetitions:Int, param:String) {
        self.init()
        if repetitions<at {self.at=repetitions}
        self.param=param
    }
    override func action(ear: String, skin: String, eye: String) -> String {
        var axnStr=""
        if self.at>0{
            if ear.lowercased() != self.param{
                axnStr=self.param
                self.at -= 1
            }
        }
        return axnStr
    }
    override func completed() -> Bool {
        return self.at < 1
    }
}

class APVerbatim: AlgPart {
    private var buffer: [String]
    private var headIndex: Int = 0
    
    init(_ sentences: String...) {
        self.buffer = sentences
    }
    
    init(_ list1: [String]) {
        self.buffer = list1
    }
    
    @inline(__always)
    override func action(ear: String, skin: String, eye: String) -> String {
        // Branchless O(1) dequeue
        let result = headIndex < buffer.count ? buffer[headIndex] : ""
        headIndex &+= 1  // Overflow-safe (unlikely but correct)
        return result
    }
    
    @inline(__always)
    override func completed() -> Bool {
        // Single branch (compiler optimizes this aggressively)
        headIndex >= buffer.count
    }
    
    /// Optional: Manual memory cleanup for very large queues
    func compact() {
        if headIndex > 1024 {  // Only reclaim memory after significant consumption
            buffer.removeFirst(headIndex)
            headIndex = 0
        }
    }
}

// a step by step plan to achieve a goal
class Algorithm{
    var algParts: Array<AlgPart> = [AlgPart]()
    init(algParts: Array<AlgPart>) {
        self.algParts = algParts
    }
    init(_ algParts: AlgPart...) {
        self.algParts = algParts
    }
    func getSize() -> Int {
        return algParts.count
    }
}
class Kokoro{
    private var emot:String = ""
    public var grimoireMemento:AbsDictionaryDB
    var toHeart:[String:String] = [:]
    init(absDictionaryDB: AbsDictionaryDB) {
        self.grimoireMemento = absDictionaryDB
    }
    func getEmot()->String{
        return emot
    }
    func setEmot(emot:String){
        self.emot = emot
    }
}
// used to transport algorithms to other classes
class Neuron{
    private var defcons:[Int:Array<Algorithm>] = [:]
    init() {
        for i in 1...6{
            defcons[i] = [Algorithm]()
        }
    }
    func insertAlg(priority:Int, alg:Algorithm){
        if priority>0 && priority < 6 {
            if defcons[priority]!.count < 4{
                defcons[priority]!.append(alg)
            }
        }
    }
    func getAlg(defcon:Int)->Algorithm?{
        if defcons[defcon]!.count>0{
            let temp:Algorithm = defcons[defcon]!.remove(at: 0)
            return temp
        }
        return nil
    }
}
open class Skill{
    private(set) var kokoro:Kokoro? = nil
    var outAlg:Algorithm? = nil
    var outpAlgPriority:Int = -1 // defcon 1->5
    init() {}
    func input(ear:String, skin:String, eye:String){
    }
    func output(noiron:Neuron){
        if let notNilAlg = self.outAlg{
            noiron.insertAlg(priority: outpAlgPriority, alg: notNilAlg)
            self.outpAlgPriority = -1
            self.outAlg = nil
        }
    }
    func setKokoro(kokoro:Kokoro){
        // use this for telepathic communication between different chobits objects
        self.kokoro = kokoro
    }
    // in skill algorithm building shortcut methods:
    func setVerbatimAlgFromList(priority:Int,  sayThis: Array<String>) {
        // build a simple output algorithm to speak string by string per think cycle
        // uses list param
        self.outAlg = Algorithm(APVerbatim(sayThis))
        self.outpAlgPriority = priority // 1->5, 1 is the highest algorithm priority
    }
    func setVerbatimAlg(priority:Int,  sayThis:String...) {
        // build a simple output algorithm to speak string by string per think cycle
        // uses varargs param
        self.outAlg = Algorithm(APVerbatim(sayThis))
        self.outpAlgPriority = priority // 1->5, 1 is the highest algorithm priority
    }
    func setSimpleAlg(sayThis:String...) {
        // based on the setVerbatimAlg method
        // build a simple output algorithm to speak string by string per think cycle
        // uses varargs param
        self.outAlg = Algorithm(APVerbatim(sayThis))
        self.outpAlgPriority = 4 // Default priority of 4
    }
    func algPartsFusion(priority:Int, algParts: AlgPart...) {
        self.outAlg = Algorithm(algParts: algParts)
        self.outpAlgPriority = priority // 1->5, 1 is the highest algorithm priority
    }
    func pendingAlgorithm() -> Bool {
        return outAlg != nil
    }
    
    func skillNotes(param: String) -> String {
        return "notes unknown"
    }
}
class DiHelloWorld:Skill{
    // hello world skill for testing purposes
    override func input(ear: String, skin: String, eye: String) {
        switch (ear)  {
          case "hello":
            super.setVerbatimAlg(priority: 4, sayThis: "hello world")
          case "incantation 0":
            // cancel running algorithm entirely at any alg part point
            super.setVerbatimAlg(priority: 4, sayThis: "fly","bless of magic caster","infinity wall", "magic ward holy","life essence")

        default:
            return
        }
    }
    override func skillNotes(param: String) -> String {
        if param == "notes" {
            return "plain hello world skill"
        } else if param == "triggers" {
            return "say hello"
        }
        return "note unavailable"
    }
}
class Cerabellum{
    // runs an algorithm
    private var fin:Int = 0
    private(set) var at:Int = 0
    private var incrementAt:Bool = false
    var alg:Algorithm?
    private var isActive1:Bool = false
    private(set) var emot:String = ""
    func advanceInAlg() {
        if incrementAt {
            incrementAt = false
            at += 1
            if at == fin {
                isActive1 = false
            }
        }
    }
    func setAlgorithm(algorithm:Algorithm) {
        if(!isActive1) && (!algorithm.algParts.isEmpty){
            self.alg = algorithm
            at = 0;fin = algorithm.getSize();isActive1 = true
            emot = alg!.algParts[at].myName()
        }
    }
    func isActive() -> Bool {
        return isActive1
    }
    func act(ear: String, skin: String, eye: String) -> String {
        if !isActive1 {return ""}
        var axnStr:String = ""
        if at < fin {
        let algPart:AlgPart = alg!.algParts[at]
            axnStr = algPart.action(ear: ear, skin: skin, eye: eye)
            self.emot = algPart.myName()
            if algPart.completed(){
                incrementAt = true
            }
        }
        return axnStr
    }
    func deActivation() {
        self.isActive1 = self.isActive1 && !alg!.algParts[at].algKillSwitch
    }
}
class Fusion {
    private var emot:String = "" // emotion represented by the active alg part (Mutatable)
    private var result:String = ""
    private var CeraArr = [Cerabellum(),Cerabellum(),Cerabellum(),Cerabellum(),Cerabellum()]
    func getEmot() -> String {
        return emot
    }
    func loadAlgs(neuron:Neuron) {
        for i in 1...5{
            if !CeraArr[i-1].isActive(){
                if let temp:Algorithm = neuron.getAlg(defcon: i){
                    CeraArr[i-1].setAlgorithm(algorithm: temp)
                }
            }
        }
    }
    func runAlgs(ear: String, skin: String, eye: String) -> String {
        result = ""
        for i in 0...4 {
            if !CeraArr[i].isActive(){continue}
            result = CeraArr[i].act(ear: ear, skin: skin, eye: eye)
            CeraArr[i].advanceInAlg()
            emot = CeraArr[i].emot
            CeraArr[i].deActivation() // deactivation if Mutatable.algkillswitch = true
            return result
        }
        emot = ""
        return result
    }
}
class Chobits {
    var dClasses: [Skill] = []
    var fusion: Fusion
    var noiron: Neuron
    var kokoro: Kokoro = Kokoro(absDictionaryDB: AbsDictionaryDB()) // consciousness
    private var isThinking: Bool = false
    private var awareSkills: [Skill] = []
    var algTriggered: Bool = false
    var ctsSkills: [Skill] = []  // continuous skills
    
    init() {
        self.fusion = Fusion()
        self.noiron = Neuron()
    }
    
    func setDataBase(absDictionaryDB: AbsDictionaryDB) {
        self.kokoro.grimoireMemento = absDictionaryDB
    }
    
    @discardableResult
    func addSkill(skill: Skill) -> Chobits {
        // add a skill (builder design patterned func)
        if self.isThinking {
            return self
        }
        skill.setKokoro(kokoro: self.kokoro)
        self.dClasses.append(skill)
        return self
    }
    
    func addSkillAware(skill: Skill){
        // add a skill with Chobit Object in their constructor
        skill.setKokoro(kokoro: self.kokoro)
        self.awareSkills.append(skill)
    }
    
    func clearSkills() {
        // remove all skills
        if self.isThinking {
            return
        }
        self.dClasses.removeAll()
    }
    
    func addSkills(skills: Skill...) {
        if self.isThinking {
            return
        }
        for skill in skills {
            skill.setKokoro(kokoro: self.kokoro)
            self.dClasses.append(skill)
        }
    }
    
    func removeSkill(skill: Skill) {
        if self.isThinking {
            return
        }
        if let index = self.dClasses.firstIndex(where: { $0 === skill }) {
            self.dClasses.remove(at: index)
        }
    }
    
    func containsSkill(skill: Skill) -> Bool {
        return self.dClasses.contains(where: { $0 === skill })
    }
    @discardableResult
    func think(ear: String, skin: String, eye: String) -> String {
        algTriggered = false
        self.isThinking = true
        for dCls in self.dClasses {
            inOut(dClass: dCls, ear: ear, skin: skin, eye: eye)
        }
        self.isThinking = false
        for dCls2 in self.awareSkills {
            inOut(dClass: dCls2, ear: ear, skin: skin, eye: eye)
        }
        isThinking = true
        for dCls3 in ctsSkills {
            if algTriggered { break }
            inOut(dClass: dCls3, ear: ear, skin: skin, eye: eye)
        }
        isThinking = false
        fusion.loadAlgs(neuron: noiron)
        return fusion.runAlgs(ear: ear, skin: skin, eye: eye)
    }
    
    func getSoulEmotion() -> String {
        return fusion.getEmot()
    }
    
    private func inOut(dClass: Skill, ear: String, skin: String, eye: String) {
        dClass.input(ear: ear, skin: skin, eye: eye)
        if dClass.pendingAlgorithm() { algTriggered = true }
        dClass.output(noiron: noiron)
    }
    
    func getKokoro() -> Kokoro {
        return kokoro
    }
    
    func setKokoro(kokoro: Kokoro) {
        self.kokoro = kokoro
    }
    
    func getFusion() -> Fusion {
        return fusion
    }
    
    func getSkillList() -> [String] {
        var result: [String] = []
        for skill in self.dClasses {
            result.append(String(describing: type(of: skill)))
        }
        return result
    }
    func addContinuousSkill(skill: Skill) {
        if isThinking { return }
        skill.setKokoro(kokoro: kokoro)
        ctsSkills.append(skill)
    }
    func clearContinuousSkills() {
        if isThinking { return }
        ctsSkills.removeAll()
    }
    func removeContinuousSkill(skill: Skill) {
        if isThinking { return }
        ctsSkills.removeAll { $0 === skill }
    }
}

public class Brain {
    var logicChobit = Chobits()
    private var emotion = ""
    private var logicChobitOutput = ""
    var hardwareChobit = Chobits()
    var ear = Chobits()
    var skin = Chobits()
    var eye = Chobits()
    // ret active alg part representing emotion
    public func getEmotion() -> String {
        return emotion
    }
    // ret feedback (last output)
    public func getLogicChobitOutput() -> String {
        return logicChobitOutput
    }
    // c'tor
    public init() {
        Brain.imprintSoul(kokoro: logicChobit.getKokoro(), args: hardwareChobit, ear, skin, eye)
    }

    static func imprintSoul(kokoro: Kokoro, args: Chobits...) {
        for arg in args {
            arg.setKokoro(kokoro: kokoro)
        }
    }
    // live
    public func doIt(ear: String, skin: String, eye: String) {
        logicChobitOutput = logicChobit.think(ear: ear, skin: skin, eye: eye)
        emotion = logicChobit.getSoulEmotion()
        hardwareChobit.think(ear: logicChobitOutput, skin: skin, eye: eye)
    }
    // add regular thinking(logical) skill
    public func addLogicalSkill(skill: Skill) {
        logicChobit.addSkill(skill: skill)
    }
    // add output skill
    public func addHardwareSkill(skill: Skill) {
        hardwareChobit.addSkill(skill: skill)
    }

    // add visual input skill
    public func addEarSkill(skill: Skill) {
        ear.addSkill(skill: skill)
    }
    // add sensor input skill
    public func addSkinSkill(skill: Skill) {
        skin.addSkill(skill: skill)
    }
    // add visual input skill
    public func addEyeSkill(skill: Skill) {
        eye.addSkill(skill: skill)
    }

    public func think(_ keyIn: String) {
        if !keyIn.isEmpty {
            // handles typed inputs(keyIn)
            doIt(ear: keyIn, skin: "", eye: "")
        } else {
            // accounts for sensory inputs
            doIt(ear: self.ear.think(ear: "", skin: "", eye: ""),
                 skin: skin.think(ear: "", skin: "", eye: ""),
                 eye: eye.think(ear: "", skin: "", eye: ""))
        }
    }

    public func think() {
        doIt(ear: ear.think(ear: "", skin: "", eye: ""),
             skin: skin.think(ear: "", skin: "", eye: ""),
             eye: eye.think(ear: "", skin: "", eye: ""))
    }
}

class DiSysOut:Skill{
    // hello world skill for testing purposes
    override func input(ear: String, skin: String, eye: String) {
        if(!ear.isEmpty && !ear.contains("#")){
            print(ear)
        }
    }
    override func skillNotes(param: String) -> String {
        if param == "notes" {
            return "prints to console"
        } else if param == "triggers" {
            return "automatic for any input"
        }
        return "note unavailable"
    }
}

File: livingrimoire start here/LivinGrimoire C++/split C++ file version/livingrimoire.cpp
#include "LivinGrimoire.h"

//APSay:Mutatable
APSay::APSay(int repetitions, const std::string& szParam) : AlgPart(), param(szParam) {
	if (repetitions > 10) {
		repetitions = 10;
	}
	at = repetitions;
}

bool APSay::completed() {
	return at < 1;
}

std::string APSay::action(const std::string& ear, const std::string& skin, const std::string& eye) {
	std::string axnStr = "";
	if (at > 0) {

		if (!lgUtil::iequals(ear, param)) {
			axnStr = param;
			at--;
		}
	}
	return axnStr;
}

//Algorithm
Algorithm::Algorithm(std::vector<std::shared_ptr<AlgPart>>& vecAlgParts) {
	algParts.insert(algParts.begin(), vecAlgParts.begin(), vecAlgParts.end());
}

Algorithm::Algorithm(std::initializer_list<std::shared_ptr<AlgPart>> vecAlgParts) {
	for (auto& word : vecAlgParts) {
		algParts.push_back(word);
	}
}

std::vector<std::shared_ptr<AlgPart>>& Algorithm::getAlgParts() {
	return algParts;
}

int Algorithm::getSize() {
	return static_cast<int>(algParts.size());
}

//APVerbatim:APVerbatim

APVerbatim::APVerbatim(std::initializer_list<std::string> initlist) : AlgPart() {
	for (auto& word : initlist) {
		sentences.push(word);
	}
}

APVerbatim::APVerbatim(std::vector<std::string>& list1) : AlgPart() {
	for (auto& word : list1) {
		sentences.push(word);
	}
}

std::string APVerbatim::action(const std::string& ear, const std::string& skin, const std::string& eye) {
	if (!sentences.empty()) {
		std::string szNext = sentences.front();
		sentences.pop();
		return szNext;
	}

	return "";
}

bool APVerbatim::completed() {
	return sentences.empty();
}

//Kokoro
Kokoro::Kokoro(std::shared_ptr<AbsDictionaryDB> lpAbsDictionaryDB) {
	grimoireMemento = lpAbsDictionaryDB;
}

void Kokoro::setGrimoireMemento(std::shared_ptr<AbsDictionaryDB> absDictionaryDB) {
	grimoireMemento = absDictionaryDB;
}

std::string Kokoro::getEmot() {
	return emot;
}

void Kokoro::setEmot(const std::string& szEmot) {
	emot = szEmot;
}

AbsDictionaryDB* Kokoro::getGrimoireMemento() {
	return grimoireMemento.get();
}

//Neuron
Neuron::Neuron() {}

void Neuron::insertAlg(int priority, const std::shared_ptr<Algorithm> alg) {
	defcons[priority].push(alg);
}

std::shared_ptr<Algorithm> Neuron::getAlg(int defcon) {
	if (defcons.at(defcon).empty())
		return nullptr;
	std::shared_ptr<Algorithm> alg = defcons.at(defcon).front();
	defcons.at(defcon).pop();
	return alg;
}

//Skill
bool Skill::pendingAlgorithm() {
	// is an algorithm pending?
	return outAlg != nullptr;
}

// in skill algorithm building shortcut methods:
void Skill::setVerbatimAlg(int priority, std::initializer_list<std::string> sayThis) {
	// build a simple output algorithm to speak std::string by std::string per think cycle
	// uses varargs param
	std::shared_ptr<APVerbatim> mut = std::make_shared<APVerbatim>(sayThis);
	std::initializer_list<std::shared_ptr<AlgPart>> list = { mut };
	outAlg = std::make_shared<Algorithm>(list);
	outpAlgPriority = priority; // 1->5 1 is the highest algorithm priority
}

void Skill::setSimpleAlg(std::initializer_list<std::string> sayThis) {
	// based on the setVerbatimAlg method
	// build a simple output algorithm to speak std::string by std::string per think cycle
	// uses varargs param
	std::shared_ptr<APVerbatim> mut = std::make_shared<APVerbatim>(sayThis);
	std::initializer_list<std::shared_ptr<AlgPart>> list = { mut };
	outAlg = std::make_shared<Algorithm>(list);
	outpAlgPriority = 4; // 1->5 1 is the highest algorithm priority
}

void Skill::setVerbatimAlgFromList(int priority, std::vector<std::string> sayThis) {
	// build a simple output algorithm to speak std::string by std::string per think cycle
	// uses list param
	std::shared_ptr<APVerbatim> mut = std::make_shared<APVerbatim>(sayThis);
	std::initializer_list<std::shared_ptr<AlgPart>> list = { mut };
	outAlg = std::make_shared<Algorithm>(list);
	outpAlgPriority = priority; // 1->5 1 is the highest algorithm priority
}

void Skill::algPartsFusion(int priority, std::initializer_list<std::shared_ptr<AlgPart>> algParts) {
	// build a custom algorithm out of a chain of algorithm parts(actions)
	outAlg = std::make_shared<Algorithm>(algParts);
	outpAlgPriority = priority; // 1->5 1 is the highest algorithm priority
}

// extraction of skill algorithm to run (if there is one)
void Skill::output(Neuron* neuron) {
	if (outAlg != nullptr) {
		neuron->insertAlg(outpAlgPriority, outAlg);
		outpAlgPriority = -1;
		outAlg = nullptr;
	}
}

void Skill::setKokoro(Kokoro* lpKokoro) {
	// use this for telepathic communication between different chobits objects
	kokoro = lpKokoro;
}

std::string Skill::skillNotes(std::string& param) {
	return "notes unknown";
}


//DiSysOut
void DiSysOut::input(const std::string& ear, const std::string& skin, const std::string& eye) {
	if (!ear.empty() && ear.find("#") == std::string::npos) {
		std::cout << ear << std::endl;
	}
}

std::string DiSysOut::skillNotes(std::string& param) {
	if (param == "notes") {
		return "console print skill";
	}
	else if ("triggers" == param) {
		return "prints any string input";
	}
	return "note unavailable";
}

//DiHelloWorld:Skill
void DiHelloWorld::input(const std::string& ear, const std::string& skin, const std::string& eye) {
	if (ear == "hello") {
		setVerbatimAlg(4, { "hello world" }); // 1->5 1 is the highest algorithm priority
	}
}

std::string DiHelloWorld::skillNotes(std::string& param) {
	if (param == "notes") {
		return "plain hello world skill";
	}
	else if ("triggers" == param) {
		return "say hello";
	}
	return "note unavailable";
}

//Cerabellum
int Cerabellum::getAt() {
	return at;
}

void Cerabellum::advanceInAlg() {
	if (incrementAt) {
		incrementAt = false;
		at++;
		if (at == fin) {
			bIsActive = false;
		}
	}
}

std::string Cerabellum::getEmot() {
	return emot;
}

void Cerabellum::setAlgorithm(const std::shared_ptr<Algorithm> algorithm) {
	if (!bIsActive && (!algorithm->getAlgParts().empty())) {
		alg = algorithm;
		at = 0;
		fin = algorithm->getSize();
		bIsActive = true;
		emot = alg->getAlgParts().at(at)->myName(); // updated line
	}
}

bool Cerabellum::isActive() {
	return bIsActive;
}

std::string Cerabellum::act(const std::string& ear, const std::string& skin, const std::string& eye) {
	std::string axnStr = "";
	if (!bIsActive) {
		return axnStr;
	}
	if (at < fin) {
		std::shared_ptr<AlgPart> mut = alg->getAlgParts().at(at);
		axnStr = mut->action(ear, skin, eye);
		emot = alg->getAlgParts().at(at)->myName();
		if (alg->getAlgParts().at(at)->completed()) {
			incrementAt = true;
		}
	}
	return axnStr;
}

void Cerabellum::deActivation() {
	bIsActive = bIsActive && !alg->getAlgParts().at(at)->algKillSwitch;
}

//Fusion
Fusion::Fusion() {}

std::string Fusion::getEmot() {
	return emot;
}

void Fusion::loadAlgs(Neuron* neuron) {
	for (int i = 0; i < ceraArr.size(); i++) {
		if (!ceraArr[i].isActive()) {
			std::shared_ptr<Algorithm> temp = neuron->getAlg(i);
			if (temp != nullptr) {
				ceraArr[i].setAlgorithm(temp);
			}
		}
	}
}

std::string Fusion::runAlgs(const std::string& ear, const std::string& skin, const std::string& eye) {
	std::string result;
	for (int i = 0; i < 5; i++) {
		if (!ceraArr[i].isActive()) {
			continue;
		}
		result = ceraArr[i].act(ear, skin, eye);
		ceraArr[i].advanceInAlg();
		emot = ceraArr[i].getEmot();
		ceraArr[i].deActivation(); // deactivation if Mutatable.algkillswitch = true
		return result;
	}
	emot.clear();
	return result;
}

//Chobits
Chobits::Chobits() : isThinking(false),algTriggered(false), fusion(std::make_unique<Fusion>()), neuron(std::make_unique<Neuron>()) {
	kokoro = std::make_shared<Kokoro>(std::make_shared<AbsDictionaryDB>());
}

Chobits::~Chobits() {
	clearSkills();
}

void Chobits::setDataBase(std::shared_ptr<AbsDictionaryDB> absDictionaryDB) {
	kokoro->setGrimoireMemento(absDictionaryDB);
}

Chobits* Chobits::addSkill(Skill* skill) {
	// add a skill (builder design patterned func))
	if (isThinking) return this;
	skill->setKokoro(kokoro.get());
	dClasses.push_back(skill);
	return this;
}

void Chobits::addContinuousSkill(Skill* skill) {
	skill->setKokoro(kokoro.get());
	ctsSkills.push_back(skill);
}

void Chobits::addSkillAware(Skill* skill) {
	skill->setKokoro(kokoro.get());
	awareSkills.push_back(skill);
}

void Chobits::clearSkills() {
	if (isThinking) return;

	for_each(dClasses.begin(), dClasses.end(), [](Skill* lpSkill) { delete lpSkill; });
	dClasses.clear();
}

void Chobits::clearContinuousSkills() {
	if (isThinking) return;

	for_each(ctsSkills.begin(), ctsSkills.end(), [](Skill* lpSkill) { delete lpSkill; });
	ctsSkills.clear();
}

void Chobits::addSkills(std::initializer_list<Skill*> skills) {
	if (isThinking) return;

	for (auto skill : skills) {
		skill->setKokoro(kokoro.get());
		dClasses.push_back(skill);
	}
}

void Chobits::removeSkill(Skill* skill) {
	if (isThinking) return;
	if (skill != nullptr) {
		auto it = find(dClasses.begin(), dClasses.end(), skill);
		if (it != dClasses.end())
			delete* it;
		dClasses.erase(it);
	}
}

void Chobits::removeContinuousSkill(Skill* skill) {
	if (isThinking) return;
	if (skill != nullptr) {
		auto it = find(ctsSkills.begin(), ctsSkills.end(), skill);
		if (it != ctsSkills.end())
			delete* it;
		ctsSkills.erase(it);
	}
}

bool Chobits::containsSkill(Skill* skill) {
	return (skill != nullptr && find(dClasses.begin(), dClasses.end(), skill) != dClasses.end());
}

std::string Chobits::think(const std::string& ear, const std::string& skin, const std::string& eye) {
	algTriggered = false;
	isThinking = true;
	for (Skill* dCls : dClasses) {
		inOut(dCls, ear, skin, eye);
	}
	isThinking = false;
	for (Skill* dCls : awareSkills) {
		inOut(dCls, ear, skin, eye);
	}
	isThinking = true;
	for (Skill* dCls : ctsSkills) {
		inOut(dCls, ear, skin, eye);
	}
	isThinking = false;
	fusion->loadAlgs(neuron.get());
	return fusion->runAlgs(ear, skin, eye);
}

std::string Chobits::getSoulEmotion() {
	return fusion->getEmot();
}

std::shared_ptr<Kokoro> Chobits::getKokoro() {
	return kokoro;
}

void Chobits::setKokoro(std::shared_ptr<Kokoro> lpKokoro) {
	kokoro = lpKokoro;
}


Fusion* Chobits::getFusion() {
	return fusion.get();
}

std::vector<std::string> Chobits::getSkillList(std::vector<std::string>& result) {
	for (Skill* skill : dClasses) {
		result.push_back(skill->myName());
	}
	return result;
}

std::string Skill::myName() {
	// Returns the class name
	return typeid(this).name();
}

void Chobits::inOut(Skill* dClass, const std::string& ear, const std::string& skin, const std::string& eye) {
	dClass->input(ear, skin, eye);
	if (dClass->pendingAlgorithm()) { algTriggered = true; }
	dClass->output(neuron.get());
}

//Brain
Brain::Brain() {
	logicChobit = std::make_unique<Chobits>();
	hardwareChobit = std::make_unique<Chobits>();
	ear = std::make_unique<Chobits>();
	skin = std::make_unique<Chobits>();
	eye = std::make_unique<Chobits>();
	imprintSoul(logicChobit->getKokoro(), { hardwareChobit.get(), ear.get(), skin.get(), eye.get() });
}

Chobits* Brain::getLogicChobit() {
	return logicChobit.get();
}

Chobits* Brain::getHardwareChobit() {
	return hardwareChobit.get();
}

std::string Brain::getEmotion() {
	return emotion;
}

std::string Brain::getLogicChobitOutput() {
	return logicChobitOutput;
}

void Brain::doIt(const std::string& ear, const std::string& skin, const std::string& eye) {
	logicChobitOutput = logicChobit->think(ear, skin, eye);
	emotion = logicChobit->getSoulEmotion();
	// case: hardware skill wishes to pass info to logical chobit
	hardwareChobit->think(logicChobitOutput, skin, eye);
}

void Brain::addLogicalSkill(Skill* skill) {
	logicChobit->addSkill(skill);
}

void Brain::addHardwareSkill(Skill* skill) {
	hardwareChobit->addSkill(skill);
}

void Brain::addEarSkill(Skill* skill) {
	ear->addSkill(skill);
}

void Brain::addSkinSkill(Skill* skill) {
	skin->addSkill(skill);
}

void Brain::addEyeSkill(Skill* skill) {
	eye->addSkill(skill);
}

void Brain::think() {
	doIt(ear->think("", "", ""), skin->think("", "", ""), eye->think("", "", ""));
}

void Brain::think(const std::string& keyIn) {
	if (!keyIn.empty()) {
		// handles typed inputs(keyIn)
		doIt(keyIn, "", "");
	}
	else {
		// accounts for sensory inputs
		doIt(ear->think("", "", ""), skin->think("", "", ""), eye->think("", "", ""));
	}
}

void Brain::imprintSoul(std::shared_ptr<Kokoro> kokoro, std::initializer_list<Chobits*> args) {
	for (auto arg : args) {
		arg->setKokoro(kokoro);
	}
}